### DISCUSSION TITLE
SWR is not passing the cookies to the server-side rendering

### DISCUSSION THREAD

**Discussion by alnow0021** [nuxt/nuxt#20014] [Questions]
if a route is cached by SWR, then there is no way to read a cookie value via `useCookie` function (on the server side, on the client side it's ok), on the server side it always returns `undefined`, if the route is not cached by SWR, then there is no such problem. Is it a bug or it have to work in this way? Or maybe you know a workaround to get a cookie value correctly on the server side

e.g.:
`nuxt.config.ts`

```ts
routeRules: {
  '/cached-page': { swr: 120 },
  '/not-cached-page': { cache: false }
},
```

a cookie is read in my plugin in this way:
```js
const myCookie= useCookie('myCookie')
console.log(myCookie) // /cached-page -> undefined
console.log(myCookie) // /not-cached-page -> [a_cookie_value]
```

summary: 
- under the page `/cached-page` there is no way to read a cookie on the server side
- under the page `/not-cached-page` I can read a cookie on the server side

how can I fix it?

tested on versions:
- 3.1.0
- 3.3.2

**Comment by qkreltms** [comment]
Cookies are not alos set in nuxt 3.7.3 ~ 3.8. [reproduce example](https://stackblitz.com/edit/nuxt-starter-dgf4ty?file=app.vue,pages%2Findex.vue,pages%2Ftest.vue,nuxt.config.ts,package.json,.gitignore). I don't know why... is it intended behavior?

**Reply by qkreltms** [reply]
I found a discussion in nitro: https://github.com/unjs/nitro/discussions/1023. It seems emptying header is intended behavior.

**Comment by Tofandel** [comment]
This is intended behavior, because in SWR you are generating a cached version of the page, those cached pages are meant to be stateless as they will be sent to anyone who visits that page. Otherwise imagine you are using the auth cookie, a user visits the page, it gets cached by SWR and then the following users get the page of the logged in user.

Do you see how that would be a security issue with user data leaking from the cache?

By removing the cookies during cache generation, this is entirely avoided and forces you to use the proper pattern. Which is to run an additional stateful request on the frontend after it's hydrated to get the necessary data. If you absolutely need to use a cookie server side, then your page should not be SWR and only SSR.

**Comment by jibin7jose** [comment]
Building on what's already been said: yes, this is intended — Nitro strips per-request data like cookies during SWR generation to avoid leaking one user's data into a cache shared by everyone. Here are a few more concrete ways to handle it depending on your case.

## Option 1: Split the route so only the personalized part is dynamic

Instead of making the whole page SSR just because of one cookie read, keep the page itself on `swr` and pull the cookie-dependent piece out into its own server route or API endpoint that is *not* cached:

```ts
// nuxt.config.ts
routeRules: {
  '/cached-page': { swr: 120 },
  '/api/user-data': { cache: false }
}
```

Then call `/api/user-data` from the client (e.g. with `useFetch` inside `onMounted`, or `import.meta.client` guard) to get anything that depends on the cookie. The page shell stays cacheable and fast; only the small personalized call is stateful.

## Option 2: Use `event.node.req.headers.cookie` in a non-cached server middleware

If you need the cookie earlier in the pipeline than component code, a `server/middleware/*.ts` file that's excluded from the cached route (or attached to an uncached sibling route) can read `event.node.req.headers.cookie` directly and set a header/response the client then reads — but note this still can't safely run *inside* the cached render path for the same leak reason.

## Option 3: Use `cachedEventHandler` with a custom `getKey` if you're rolling your own API caching

If it's your own API route (not the page render) that you want cached but still cookie-aware, Nitro's `defineCachedEventHandler` lets you vary the cache key, e.g. by a non-sensitive user segment, so different cookie states get different cache entries rather than none at all:

```ts
export default defineCachedEventHandler((event) => {
  // ...
}, {
  maxAge: 60,
  getKey: (event) => `data-${getCookie(event, 'segment')}`
})
```

This is a narrower tool than page-level `swr` and only makes sense if the cookie value maps to a small, safe set of variants (not per-user secrets) — otherwise you're back to the same leak risk, just scoped smaller.

## Bottom line

Page-level `swr` is inherently for cacheable, shared HTML. Anything that must vary per visitor's cookie needs to either live outside the cached render (client-side fetch, or an uncached API route) or use a caching strategy that explicitly keys on the safe parts of that cookie, never the raw value.
